src/main/java/org/antlr/jetbrains/adapter/lexer/PsiElementTypeFactory.java

```diff
@@ -92,11 +100,16 @@ public TokenIElementType getEofElementType() {
     }
 
     @NotNull
-    private List<RuleIElementType> createRuleIElementTypes(Language language, String[] ruleNames) {
+    private List<RuleIElementType> createRuleIElementTypes(Language language, String[] ruleNames,
+            Collection<RuleIElementType> customRuleElementTypes) {
         List<RuleIElementType> result;
-        RuleIElementType[] elementTypes = new RuleIElementType[ruleNames.length];
+        RuleIElementType[] elementTypes = new RuleIElementTypeImpl[ruleNames.length];
         for (int i = 0; i < ruleNames.length; i++) {
-            elementTypes[i] = new RuleIElementType(i, ruleNames[i], language);
+            elementTypes[i] = new RuleIElementTypeImpl(i, ruleNames[i], language);
+        }
+        for (RuleIElementType customType : customRuleElementTypes) {
+            assert customType instanceof IElementType;
```

Comment:
Please do not use `assert`, assertions are not enabled by default (jvm should run with `-ea` flag to turn them on).
Better replace it with `Preconditions` (if we have guava as a dependency) or something similar.